python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,356 @@
+            class label of some model for all the batched graphs,
+            which is of shape :math:`(1, L)`, where :math:`L` is the
+            different types of label in the dataset.
+        ori_pred: Tensor
+            Tensor of shape ::math:`(1, 1)`, representing the original prediction
+            for the graph.
+
+        Returns
+        -------
+        float
+            The function that returns the sum of the three loss components,
+            which is a scalar tensor representing the total loss.
+        """
+        target_prob = prob.gather(-1, ori_pred.unsqueeze(-1))
+        # 1e-6 added to prob to avoid taking the logarithm of zero
+        target_prob += 1e-6
+        # computing the log likelihood for a single prediction
+        pred_loss = torch.mean(-torch.log(target_prob))
+
+        # size
+        edge_mask = self.sparse_mask_values
+        if self.coff_budget <= 0:
+            size_loss = self.coff_budget * torch.sum(edge_mask)
+        else:
+            size_loss = self.coff_budget * F.relu(
+                torch.sum(edge_mask) - self.coff_budget
+            )
+
+        # entropy
+        scale = 0.99
+        edge_mask = self.edge_mask * (2 * scale - 1.0) + (1.0 - scale)
+        mask_ent = -edge_mask * torch.log(edge_mask) - (
+            1 - edge_mask
+        ) * torch.log(1 - edge_mask)
+        mask_ent_loss = self.coff_connect * torch.mean(mask_ent)
+
+        loss = pred_loss + size_loss + mask_ent_loss
+        return loss
+
+    def concrete_sample(self, log_alpha, beta=1.0, training=True):
+        r"""Sample from the instantiation of concrete distribution when training.
+
+        Parameters
+        ----------
+        log_alpha : Tensor
+            A tensor representing the log of the prior probability of activating the gate.
+        beta : float, optional
+            Controls the degree of randomness in the gate's output.
+        training : bool, optional
+            Indicates whether the gate is being used during training or evaluation.
+
+        Returns
+        -------
+        Tensor
+            If training is set to True, the output is a tensor of probabilities that
+            represent the probability of activating the gate for each input element.
+            If training is set to False, the output is also a tensor of probabilities,
+            but they are determined solely by the log_alpha values, without adding any
+            random noise.
+        """
+        if training:
+            bias = self.sample_bias
+            random_noise = torch.rand(log_alpha.size()).to(log_alpha.device)
+            random_noise = bias + (1 - 2 * bias) * random_noise
+            gate_inputs = torch.log(random_noise) - torch.log(
+                1.0 - random_noise
+            )
+            gate_inputs = (gate_inputs + log_alpha) / beta
+            gate_inputs = torch.sigmoid(gate_inputs)
+        else:
+            gate_inputs = torch.sigmoid(log_alpha)
```

Comment:
Is there a source for doing this after training? A paper paragraph or a code snippet.

Reply:
https://github.com/flyingdoog/PGExplainer/blob/master/codes/Explainer.py#L56

From the original implementation.